File: python/assign_cookies.py
class Solution:
    def findContentChildren(self, g: list[int], s: list[int]) -> int:
        """time complexity: O(nlogn + mlogm)"""
        g.sort()
        s.sort()
        i: int = 0
        j: int = 0
        while i < len(g):
            while j < len(s) and g[i] > s[j]:
                j += 1
            if j == len(s):
                break
            i += 1
            j += 1
        return i


def test_findContentChildren_case_1():
    # arrange
    g: list[int] = [1, 2, 3]
    s: list[int] = [1, 1]
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.findContentChildren(g, s)

    # assert
    assert expected == actual


def test_findContentChildren_case_2():
    # arrange
    g: list[int] = [1, 2]
    s: list[int] = [1, 2, 3]
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.findContentChildren(g, s)

    # assert
    assert expected == actual

File: rust/src/code/assign_cookies.rs
struct Solution;

impl Solution {
    pub fn find_content_children(mut g: Vec<i32>, mut s: Vec<i32>) -> i32 {
        // NOTE: time complexity: O(nlogn + mlogm)
        g.sort();
        s.sort();
        let mut i: usize = 0;
        let mut j: usize = 0;
        while i < g.len() {
            while j < s.len() && g[i] > s[j] {
                j += 1;
            }
            if j == s.len() {
                break;
            }
            i += 1;
            j += 1;
        }
        i as i32
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_find_content_children_case_1() {
        // arrange
        let g: Vec<i32> = vec![1, 2, 3];
        let s: Vec<i32> = vec![1, 1];
        let expected = 1;

        // act
        let actual = Solution::find_content_children(g, s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_find_content_children_case_2() {
        // arrange
        let g: Vec<i32> = vec![1, 2];
        let s: Vec<i32> = vec![1, 2, 3];
        let expected = 2;

        // act
        let actual = Solution::find_content_children(g, s);

        // assert
        assert_eq!(actual, expected);
    }
}
